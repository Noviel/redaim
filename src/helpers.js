import { EXEC_UNKNOWN } from './constants';
/*
  Extract matching field from every object in an array or 
  subfield from every object's field and put it to object.
*/
export const extractValues = (collection = {}, field) => {
  if (typeof field !== 'string') {
    throw new Error('name should be a string');
  }
  
  return Object.keys(collection).reduce((acc, curr) => {
    const prop = collection[curr][field];
    
    if (prop !== undefined) { 
      acc[curr] = prop; 
    }

    return acc;
  }, {});
};

export const execSelector = ({ meta : { exec = EXEC_UNKNOWN } = {} }) =>  exec;

export const createReducerWithHandlers = (handlers = {}, keySelector = execSelector) => (state, action) => {
  if (typeof handlers !== 'object') {
    return state;
  }
  
  const key = keySelector(action);

  if (!handlers.hasOwnProperty(key)) {
    return state;
  }

  return handlers[key](state, action);
};
