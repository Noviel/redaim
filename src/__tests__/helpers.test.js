import { extractValues, execSelector, createReducerWithHandlers } from '../helpers';

describe('extractValues', () => {

  it('should throw with not a string field', () => {
    expect(() => extractValues([], null))
      .toThrow();
  });

  it('should return an empty object if property is not exist', () => {
    const initial = [{ a: 1, b: { a: 2 }}, { a: 'second', z: 0 }];
    expect(extractValues(initial, 'nonExisted'))
      .toMatchSnapshot();
  });

  it('should properly extract fields from an array with empty objects', () => {
    const initial = [
      {},
      { a: 1, b: { a: 2 }}, 
      { a: 'second', z: 0 },
      {}
    ];
    expect(extractValues(initial, 'a'))
      .toMatchSnapshot();
  });

  it('should properly extract fields from an array', () => {
    const initial = [
      { a: 1, b: { a: 2 }}, 
      { a: 'second', z: 0 }
    ];
    expect(extractValues(initial, 'a'))
      .toMatchSnapshot();
  });

  it('should extract properties from an object', () => {
    const initial = {
      obj1: { a: 'from obj1', c: 2},
      obj2: { b: 4 },
      obj3: { a: 'from obj3'}
    };
    expect(extractValues(initial, 'a'))
      .toMatchSnapshot();
  });
});

describe('execSelector', () => {    
  const action = {
    meta: {
      exec: 'EXEC_VALUE'
    }
  };

  it('should properly select a key', () => {
    expect(execSelector(action))
      .toMatchSnapshot();
  });

  it('should return specific error string if key is not found in the action\'s meta', () => {
    const result = execSelector({
      meta: { }
    });

    expect(result)
      .toMatchSnapshot();
  });

  it('should return specific error string if meta is not found in an action', () => {
    const result = execSelector({ 
      type: 'TYPE', 
      payload: { value: 1 }
    });

    expect(result)
      .toMatchSnapshot();
  });
});

describe('createReducerWithHandlers', () => {
  const initialState = 'initalState';

  const actionOne = {
    meta: {
      exec: 'ACTION_1'
    }
  };

  const handlers = {
    'ACTION_0': (state, action) => 'one',
    'ACTION_1': (state, action) => 'two'
  };

  it('should return unmodified state if empty handlers', () => {
    const reducer = createReducerWithHandlers();

    expect(reducer(initialState, actionOne))
      .toMatchSnapshot();
  });

  it('should return unmodified state if handlers is not an object', () => {
    const reducer = createReducerWithHandlers(42);

    expect(reducer(initialState, actionOne))
      .toMatchSnapshot();
  });
  
  it('should properly take a key from an action with default keySelector', () => {
    const reducer = createReducerWithHandlers(handlers);

    expect(reducer(initialState, actionOne))
      .toMatchSnapshot();
  });

  it('should not modify the state if key is not found in handlers', () => {
    const reducer = createReducerWithHandlers(handlers);
    const actionNotMatch = {
      meta: {
        exec: 'ACTION_UNKNOWN'
      }
    };

    expect(reducer(initialState, actionNotMatch))
      .toMatchSnapshot();
  });

  it('should invoke correct handler with custom keySelector', () => {
    const keySelector = action => action.type;
    const reducer = createReducerWithHandlers(handlers, keySelector);

    expect(reducer(initialState, { type: 'ACTION_0' }))
      .toMatchSnapshot();
  });
});
