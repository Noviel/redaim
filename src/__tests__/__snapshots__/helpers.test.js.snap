// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createReducerWithHandlers should invoke correct handler with custom keySelector 1`] = `"one"`;

exports[`createReducerWithHandlers should not modify the state if key is not found in handlers 1`] = `"initalState"`;

exports[`createReducerWithHandlers should properly take a key from an action with default keySelector 1`] = `"two"`;

exports[`createReducerWithHandlers should return unmodified state if empty handlers 1`] = `"initalState"`;

exports[`createReducerWithHandlers should return unmodified state if handlers is not an object 1`] = `"initalState"`;

exports[`execSelector should properly select a key 1`] = `"EXEC_VALUE"`;

exports[`execSelector should return specific error string if key is not found in the action's meta 1`] = `"@@REDAIM-ACTION-UNKNOWN"`;

exports[`execSelector should return specific error string if meta is not found in an action 1`] = `"@@REDAIM-ACTION-UNKNOWN"`;

exports[`extractValues should extract properties from an object 1`] = `
Object {
  "obj1": "from obj1",
  "obj3": "from obj3",
}
`;

exports[`extractValues should properly extract fields from an array 1`] = `
Object {
  "0": 1,
  "1": "second",
}
`;

exports[`extractValues should properly extract fields from an array with empty objects 1`] = `
Object {
  "1": 1,
  "2": "second",
}
`;

exports[`extractValues should return an empty object if property is not exist 1`] = `Object {}`;
